add .overrideDerivation and .overrideAttrs to packages created with callPackages

danbst opened the pull request:
###### Motivation for this change

After I've change tomcat to `callPackages` in https://github.com/NixOS/nixpkgs/pull/18419 , I've noticed that `pkgs.tomcat8.overrideDerivation` attribute is lost. This PR is my (and @aneeshusa) vision on how to restore it.
###### Things done

[x] tested override and overrideDerivation for tomcat
[ ] tested everything else

aneeshusa commented on lib/customisation.nix:
```diff
@@ -106,9 +107,7 @@ rec {
-      mkAttrOverridable = name: pkg: pkg // {
-        override = newArgs: mkAttrOverridable name (f (finalArgs // newArgs)).${name};
-      };
+      mkAttrOverridable = name: pkg: makeOverridableGeneric (x: x."${name}") f finalArgs;
```
It looks like `postProcess` is always applied directly to the output of `f` in `makeOverrideableGeneric`, so why not just pass the composition of `f` and `x: x."${name}"` to `makeOverrideable` here?

danbst replied:
my bad, how had I missed that?..

aneeshusa commented on lib/customisation.nix:
```diff
@@ -106,9 +106,7 @@ rec {
-      mkAttrOverridable = name: pkg: pkg // {
-        override = newArgs: mkAttrOverridable name (f (finalArgs // newArgs)).${name};
-      };
+      mkAttrOverridable = name: pkg: makeOverridable (newArgs: (f newArgs).${name}) finalArgs;
```
Can you also rename `finalArgs` to `origArgs`? I think that is less confusing.

danbst replied:
No problem. There are `args`, `auto`, `autoArgs`, `finalArgs` and `newArgs` in one function! So in any case it is better to follow binding declarations than relying on mnemonics.

danbst commented:
ping @edolstra

domenkozar commented:
Instead we should rather add `overrideAttrs` since that's preferred now.

Commit:
add .overrideDerivation and .overrideAttrs to packages created with `callPackages`/`callPackagesWith`

nix/nixUnstable, tomcatN and postgresqlNN use `callPackages` pattern, they have .override
attribute, but lack .overrideDerivation and recent .overrideAttrs.
Packages created with `callPackage` have all of those. Because .overrideDerivation function
is used in public, without this we can break code when refactoring callPackage -> callPackages.

danbst commented:
Rebased to master, added motivation for a change to commit message, changed topic name, tested using `.overrideAttrs`. All is good
